backend/forms/models/response.py

```diff
@@ -30,6 +32,168 @@ class Response(models.Model):
         verbose_name=_("IP address"),
     )
 
+    # denormalized fields
+    cached_dimensions = models.JSONField(
+        default=dict,
+        help_text="dimension slug -> list of value slugs",
+    )
+
+    # related fields
+    dimensions: models.QuerySet[ResponseDimensionValue]
+
+    @property
+    def survey(self) -> Survey | None:
+        return self.form.survey
+
+    def _build_cached_dimensions(self) -> dict[str, list[str]]:
+        """
+        Used by ..handlers/dimension.py to populate cached_dimensions
+        """
+        survey = self.survey
+        if survey is None:
+            return {}
+
+        new_cached_dimensions = {}
+        for sdv in self.dimensions.all():
```

Comment:
Opportunity to use `values_list("dimension__slug", "value__slug")`, I think?

Reply:
Actually it defeats the `prefetch_related` in `refresh_cached_dimensions_qs`.

With current solution:
```sql
SELECT "forms_response"."id" FROM "forms_response" FOR UPDATE OF "forms_response"
SELECT "forms_responsedimensionvalue"."id", "forms_responsedimensionvalue"."response_id", "forms_responsedimensionvalue"."dimension_id", "forms_responsedimensionvalue"."value_id" FROM "forms_responsedimensionvalue" WHERE "forms_responsedimensionvalue"."response_id" IN ('d7a9c56a78144d08ada57e0c2d88a931'::uuid)
SELECT "forms_dimension"."id", "forms_dimension"."survey_id", "forms_dimension"."slug", "forms_dimension"."title", "forms_dimension"."order", "forms_dimension"."is_key_dimension" FROM "forms_dimension" WHERE "forms_dimension"."id" IN (1, 2) ORDER BY "forms_dimension"."survey_id" ASC, "forms_dimension"."order" ASC, "forms_dimension"."slug" ASC
SELECT "forms_dimensionvalue"."id", "forms_dimensionvalue"."dimension_id", "forms_dimensionvalue"."slug", "forms_dimensionvalue"."title", "forms_dimensionvalue"."color", "forms_dimensionvalue"."is_initial" FROM "forms_dimensionvalue" WHERE "forms_dimensionvalue"."id" IN (1, 4)
UPDATE "forms_response" SET "cached_dimensions" = (CASE WHEN ("forms_response"."id" = 'd7a9c56a78144d08ada57e0c2d88a931'::uuid) THEN '{"category": ["vendor_hall"], "status": ["new"]}'::jsonb ELSE NULL END)::jsonb WHERE "forms_response"."id" IN ('d7a9c56a78144d08ada57e0c2d88a931'::uuid)
```

With the suggested `values_list`, there's an extra query number 5:

```sql
SELECT "forms_response"."id" FROM "forms_response" FOR UPDATE OF "forms_response"
SELECT "forms_responsedimensionvalue"."id", "forms_responsedimensionvalue"."response_id", "forms_responsedimensionvalue"."dimension_id", "forms_responsedimensionvalue"."value_id" FROM "forms_responsedimensionvalue" WHERE "forms_responsedimensionvalue"."response_id" IN ('d7a9c56a78144d08ada57e0c2d88a931'::uuid)
SELECT "forms_dimension"."id", "forms_dimension"."survey_id", "forms_dimension"."slug", "forms_dimension"."title", "forms_dimension"."order", "forms_dimension"."is_key_dimension" FROM "forms_dimension" WHERE "forms_dimension"."id" IN (1, 2) ORDER BY "forms_dimension"."survey_id" ASC, "forms_dimension"."order" ASC, "forms_dimension"."slug" ASC
SELECT "forms_dimensionvalue"."id", "forms_dimensionvalue"."dimension_id", "forms_dimensionvalue"."slug", "forms_dimensionvalue"."title", "forms_dimensionvalue"."color", "forms_dimensionvalue"."is_initial" FROM "forms_dimensionvalue" WHERE "forms_dimensionvalue"."id" IN (1, 4)
SELECT "forms_dimension"."slug", "forms_dimensionvalue"."slug" FROM "forms_responsedimensionvalue" INNER JOIN "forms_dimension" ON ("forms_responsedimensionvalue"."dimension_id" = "forms_dimension"."id") INNER JOIN "forms_dimensionvalue" ON ("forms_responsedimensionvalue"."value_id" = "forms_dimensionvalue"."id") WHERE "forms_responsedimensionvalue"."response_id" = 'd7a9c56a78144d08ada57e0c2d88a931'::uuid
UPDATE "forms_response" SET "cached_dimensions" = (CASE WHEN ("forms_response"."id" = 'd7a9c56a78144d08ada57e0c2d88a931'::uuid) THEN '{"category": ["vendor_hall"], "status": ["new"]}'::jsonb ELSE NULL END)::jsonb WHERE "forms_response"."id" IN ('d7a9c56a78144d08ada57e0c2d88a931'::uuid)
```